server/deviceflowhandlers_test.go

```diff
@@ -0,0 +1,672 @@
+				Expiry:     now().Add(5 * time.Minute),
+			},
+			expectedResponseCode: http.StatusBadRequest,
+		},
+		{
+			testName:     "Bad Device Request Client",
+			values:       baseFormValues,
+			testAuthCode: baseAuthCode,
+			testDeviceRequest: storage.DeviceRequest{
+				UserCode:   "XXXX-XXXX",
+				DeviceCode: "devicecode",
+				Scopes:     []string{"openid", "profile", "email"},
+				Expiry:     now().Add(5 * time.Minute),
+			},
+			expectedResponseCode: http.StatusUnauthorized,
+		},
+		{
+			testName:     "Bad Device Request Secret",
+			values:       baseFormValues,
+			testAuthCode: baseAuthCode,
+			testDeviceRequest: storage.DeviceRequest{
+				UserCode:     "XXXX-XXXX",
+				DeviceCode:   "devicecode",
+				ClientSecret: "foobar",
+				Scopes:       []string{"openid", "profile", "email"},
+				Expiry:       now().Add(5 * time.Minute),
+			},
+			expectedResponseCode: http.StatusUnauthorized,
+		},
+		{
+			testName:          "Expired Device Token",
+			values:            baseFormValues,
+			testAuthCode:      baseAuthCode,
+			testDeviceRequest: baseDeviceRequest,
+			testDeviceToken: storage.DeviceToken{
+				DeviceCode:          "devicecode",
+				Status:              deviceTokenPending,
+				Token:               "",
+				Expiry:              now().Add(-5 * time.Minute),
+				LastRequestTime:     time.Time{},
+				PollIntervalSeconds: 0,
+			},
+			expectedResponseCode: http.StatusBadRequest,
+		},
+		{
+			testName:          "Device Code Already Redeemed",
+			values:            baseFormValues,
+			testAuthCode:      baseAuthCode,
+			testDeviceRequest: baseDeviceRequest,
+			testDeviceToken: storage.DeviceToken{
+				DeviceCode:          "devicecode",
+				Status:              deviceTokenComplete,
+				Token:               "",
+				Expiry:              now().Add(5 * time.Minute),
+				LastRequestTime:     time.Time{},
+				PollIntervalSeconds: 0,
+			},
+			expectedResponseCode: http.StatusBadRequest,
+		},
+		{
+			testName:             "Successful Exchange",
+			values:               baseFormValues,
+			testAuthCode:         baseAuthCode,
+			testDeviceRequest:    baseDeviceRequest,
+			testDeviceToken:      baseDeviceToken,
+			expectedResponseCode: http.StatusOK,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.testName, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			// Setup a dex server.
+			httpServer, s := newTestServer(ctx, t, func(c *Config) {
+				//c.Issuer = c.Issuer + "/non-root-path"
+				c.Now = now
+			})
+			defer httpServer.Close()
+
+			if err := s.storage.CreateAuthCode(tc.testAuthCode); err != nil {
+				t.Errorf("failed to create auth code: %v", err)
+			}
+
+			if err := s.storage.CreateDeviceRequest(tc.testDeviceRequest); err != nil {
+				t.Errorf("failed to create device request: %v", err)
+			}
+
+			if err := s.storage.CreateDeviceToken(tc.testDeviceToken); err != nil {
+				t.Errorf("failed to create device token: %v", err)
+			}
```

Comment:
Should these be fatal errors for the test? I don't think the rest of the test is valid if these don't succeed?

Reply:
Done!